ox_bqpipeline/bqpipeline.py

```diff
@@ -297,9 +356,66 @@ def create_job_config(self, batch=False, dest=None, create=True,
                                        create_disposition=create_disp,
                                        write_disposition=write_disp)
 
+    def validate_query_params(self, query_params, depth=0):
+        """Validate the named/positional query parameters."""
+        if not (isinstance(query_params, list) or
+                isinstance(query_params, dict)):
+            return False
+        for key in query_params:
+            if isinstance(query_params, list):
+                value = key
+                key = None
+            else:
+                if not (depth or isinstance(key, str)):
+                    return False
+                value = query_params[key]
+
+            # For list, make sure each element is a scalar type.
+            # For dict, check recursively for valid parameters.
+            # for others, check that they map to valid scalar types.
+            if isinstance(value, list):
+                typ = set([type(val).__name__ for val in value])
+                res = not typ or (len(typ) == 1 and
```

Comment:
what is `not typ` handling?

Reply:
As discussed, this will not be needed as empty list will raise a ValueError upfront.

One issue that may pop up is for a use case where default value of empty list is needed for a query. Since the current approach infers the type of the QueryArrayParameter from the contents of the python list, it would not be possible to have empty lists unless an additional parameter is added to specify the BQ type in case the array is empty. I left it out to keep the parameter definition simple. Let me know if we need to reconsider this.